### ISSUE TITLE
Flag to not create a virtual root folder project?

### ISSUE THREAD

**Issue by shanon84** [khalilou88/jnxplus#1164]
hey,
maybe it is possible to add an option to not use a virtual root project?
Reasoning:
- With the virtual root project, every file in the root folder is part of a project. If you use caching, you need to consider it and not use a wildcard like "all files in project are relevant for caching except target/", but a directly "Cache src/* and pom.xml". Else you get cache relevant changes even if the node_modules folder changes for example.
- If you use access layering in frontend, the root project prevents access to the node_modules folder, if not explizitly considered.
- nx graph is bloaded cause of the relations to the root pom

We tried it out in an older jnxplus version and it worked like a charm.
We had a root pom only as an aggregator containing all project poms in modules block. Even using sub aggregator pom.xml in subfolder is possible to sort module structure without relativ path mess. A project was only created, if a project.json is next to the pom.xml.

Best regards

**Comment by khalilou88** [comment]
@shanon84 
`mavenRootDirectory` option should fix all this issues (or at least most of them). The idea is to create all maven projects in a sub folder (with config files) and you isolate maven root project from NX root project.

I don't think is a good idea to mange projects as apps and libs as in your test repo. I believe it's more interesting to manage project by technologies for better caching and performances. Inside `mavenRootDirectory` you can have two folders apps and libs if you want.

**Comment by khalilou88** [comment]
I forget to mention that adding maven root project to the graph is important to make the build target works. Check this old issue for more information https://github.com/khalilou88/jnxplus/issues/20

**Comment by shanon84** [comment]
Hey, I had a deeper look into it.
mavenRootDirectory is a nice way to prevent the frontend access rules problem. But you still need another pom.xml in root folder to automatically let the IDE detect the repository as a maven project. This one needs to still be a multi module one targeting the newly added root pom in the subfolder by referencing it as a module. (and that was my point: aggregator pom not in project referencing all the projects for the IDE, but noone use it inside the projects)
In regards of the important of the root pom built: Well, you only need this one, if you use the aggregator pom as parent in your projects. If not, you do not need to build it. I had initial problems with the root pom, because I wanted to use a different target for FE and BE. So the root pom was not installed initially :D
Another big concern with the aggregator pom + root project:
If you add a new project to your 100 maven project repository, it results in a full build/test, because your root pom changes. Even if that does not impact anything (if you dont use root pom as parent in projects).

About caching and performance problems cause of mixing the technoligies:
How does this happen? What is the (root) problem here? It would be nice to know (I would prefer layering by logic/service (FE+BE as a unit) and not technoligy based)

**Comment by khalilou88** [comment]
@shanon84 did you try to add this `pom.xml` to `.nxignore` file https://nx.dev/reference/nxignore

**Comment by khalilou88** [comment]
The problem of the aggregator pom with many modules is a common problem in Nx: here is a example with a typescript project  https://x.com/AlexOkrushko/status/1659930346169806849

**Comment by khalilou88** [comment]
In `nx-maven` we have 3 types of links between projects:
- Parent Project links
- Aggregator Project links
- Dependency links

We can add an option to remove Aggregator Project links: `skipAggregatorProjectGraphLinking` by default will be false.

To make `nx-maven` works in this case, we need to build Aggregator Projects first, then other projects.

**Comment by shanon84** [comment]
Hey,
about the caching example:
Well, one of the reasons I would like to get rid of a required aggregtor pom project ^^ without this, you dont run into huge builtload. If you do not use the aggregator pom as parent pom in any submodule, you dont need the aggregator pom at all (only as an IDE helper to automatically import the maven projects correctly).
About the option:
Well, as I tryed to explain: if you do not use your aggregator pom as a parent in a module project, you do not need to build/install the aggregator pom at all. Out of the view of the module project, there is no aggregator. This way we have only use of NX projects and dont mix it with the conzept of maven aggregator. Only while creating the dependency graph the aggregator is used to find all projects, but later it is not present at all.
Mh, in my case/idea this solution should work. But to have a broader solution: Maybe dont add a maven project as a nx project if no project.json is in the same folder? This way you can mix the usage of aggregator. The ones only for IDE helping purpose without a project representation (no project.json). And the ones you want to have with a project.json present (for, I dont know? :D )
Or are there reasons to not have a project.json, but expecting to have a project present?

**Comment by khalilou88** [comment]
I already thought about making `project.json` mandatory, but with nx project crystal, it's better to keep the file optional.

**Comment by khalilou88** [comment]
@jbadeau what do you think about `skipAggregatorProjectGraphLinking` option? is useful for you?

**Comment by jbadeau** [comment]
I think it would be useful although im not sure it will fully work. can try.

**Comment by khalilou88** [comment]
@shanon84 @jbadeau 
Please try the version 1.7.0-0.

**Comment by shanon84** [comment]
@khalilou88 
Yes, it works like planned.
Still would like the other option too to not create a project if a project.json is not present ^^
Because with skipAggregatorProjectGraphLinking active you have completely useless aggregator projects left. Could skip them to not build them/not see them in nx graph. Also not messing up root folder (if not moving the maven project path)

something like:
`  const projectJsonPath = path.join(projectAbsolutePath, 'project.json');
  if (!onlyCreateProjectWithProjectJsonPresent || fs.existsSync(projectJsonPath)) {
    projects.push({
    .....`
    
    Tested it out in my test monorepo and it worked like expected.

**Comment by khalilou88** [comment]
Maybe it will better to isolate this variables like this :
```
graph: object {
skipProjectWithoutProjectJson: string;
skipAggregatorProjectLinking : string;  // or skipAggregatorProjectLinks
}
```

**Comment by shanon84** [comment]
yes, this looks like a nice wording.

**Comment by khalilou88** [comment]
@shanon84 
Changed are published in version 1.7.2. You can use graph options like this :

```
    {
      "plugin": "@jnxplus/nx-maven",
      "options": {
        "mavenRootDirectory": "nx-maven",
        "localRepoRelativePath": ".m2/repository",
        "graphOptions": {
          "skipProjectWithoutProjectJson": true,
          "skipAggregatorProjectLinking": true
        }
      }
    }
```
I think in your case, you only need `skipProjectWithoutProjectJson`.

**Comment by shanon84** [comment]
thanks, seems to work like a charm
